Feature

EvgeniyaKuzmina opened the pull request.

EvgeniyaKuzmina committed:
feat: main-service:

- добавлен функционал по работе с комментариями. Реализован репозиторий и сервисный слой по работе с комментариями. Реализован публичный контроллер отвечающий за получение всех комментариев по указанному событию.

EvgeniyaKuzmina committed:
feat: main-service:

- доработан функционал по работе с комментариями. Реализован приватный контроллер для работы с комментариями пользователей

EvgeniyaKuzmina committed:
Merge branch 'develop_comment' into develop_with_comment

# Conflicts:
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/EventServiceImpl.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentRepository.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentService.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentServiceImpl.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentPrivateController.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentPublicController.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/dto/CommentDto.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/dto/CommentShortDto.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/dto/NewCommentDto.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/dto/UpdateCommentDto.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/mapper/CommentMapper.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/controller/EventAdminController.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/controller/EventPrivateController.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/dto/EventFullDto.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/dto/EventShortDto.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/mapper/EventMapper.java
#	main-service/src/main/java/ru/yandex/practicum/mainserver/event/controller/EventPublicController.java

EvgeniyaKuzmina committed:
Merge branch 'develop' into future

# Conflicts:
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/model/Comment.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/controller/EventPrivateController.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/dto/EventFullDto.java
#	ewm-service/src/main/java/ru/yandex/practicum/mainservice/user/UserServiceImpl.java

EvgeniyaKuzmina committed:
feat: дописана логика работы с комментариями. Администратор может либо опубликовать комментарий, либо отклонить его. Создатель комментария может изменить комментарий, после изменения комментария, статус комментария автоматически становится Pending и ожидает публикации админа.
Не авторизованные пользователи могут просмотреть все опубликованные комментарии к событию

EvgeniyaKuzmina committed:
Merge branch 'develop' into feature

EvgeniyaKuzmina committed:
feat: 1. Добавлены методы в контроллер админа для работы с комментариями. Админ может получить комментарий по id с любым статусом, или получать список всех комментариев с указанным статусом
2. Исправлены вызовы клиента и репозитория, эти методы более не вызываются внутри цикла.

EvgeniyaKuzmina committed:
feat: 1.доработаны методы для получения списка комментариев privet API и admin API. Получение всех комментариев возможно с указанием статуса, с указанием типа сортировки (убывание/возрастание по дате)

EvgeniyaKuzmina committed:
style: добавлено описание о новой функциональности приложения «комментарии»

EvgeniyaKuzmina committed:
style: удалён лишний импорт, добавлена разделительная строка

EvgeniyaKuzmina committed:
style: удалён лишний пробел, добавлены комментарии в сервис статистики

EvgeniyaKuzmina committed:
style: добавлены значения по умолчанию для полей confirmedRequest и views объекта Event

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/resources/schema.sql:
```diff
@@ -22,6 +22,7 @@ CREATE TABLE IF NOT EXISTS comments
     event_id  BIGINT                                  NOT NULL,
```
Мне кажется можно сделать event_id FOREIGN KEY на таблицу event, в этом случае данные будут согласованы между собой

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentAdminController.java:
```diff
@@ -0,0 +1,87 @@
+    private final CommentService service;
+
+    @Autowired
+    public CommentAdminController(CommentService service) {
```
Можно использовать аннотацию @RequiredArgsConstructor, она создаёт конструктор для необходимых сущностей, будет немного меньше кода

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/EventService.java:
```diff
@@ -62,6 +62,11 @@ public interface EventService {
+    /**
+     * получение списка событий по id событий
+     */
+    Collection<Event> getEventByIdIn(Collection<Long> eventId);
```
Это не замечание, хотел узнать почему решила использовать Collection<> в проекте для передачи?

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentService.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.Collection;
+
+/**
+ * класс описывающий методы для работы с комментариями
```
Комментарии не нужны. Названия классов и методов достаточно хорошо описывают что делает метод или класс. Обычно комментарии оставляют в местах, где происходит, что-то неочевидное и этого лучше избегать.

EvgeniyaKuzmina replied:
Комментарии пишу для себя, мне так удобнее. Я сужу по себе, если я буду просматривать чужую работу, мне бы хотелось видеть какие-то общие поясняющие комментарии, во избежание лишних вопросов и недопонимания

Anton-Morozov-Java-3-2 replied:
Ок, оставь для себя. 

Статья про комментарии в коде
https://habr.com/ru/company/enterra/blog/182574/

EvgeniyaKuzmina replied:
спасибо за статью)

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,185 @@
+public class CommentServiceImpl implements CommentService {
+
+    private final CommentRepository repository;
+    private final UserService userService;
```
Это моя рекомендация и мнение, необязательное к исправлению.  Лучше из сервисов обращаться к репозиториям, для того чтобы не было зависимостей между сервисами. Для небольших проектов это несущественно, но для крупных я думаю может быть важно.

EvgeniyaKuzmina replied:
Обсуждали подобный момент ранее, ещё когда я была в другой группе. И я в принципе с этим согласна и придерживаюсь такого подхода, мне он кажется логичен. Сервисы работают только со своими репозиториями. То есть сервис юреза обращается к репозиторию юзера. А если из другого сервиса  нужно что-то получить от юзера, то он обращается к сервису юзера, вызывая необходимые для него методы. 
Тот же метод получение пользователя по id, например,  — в сервисе у меня уже обработан метод findById от репозитория, вытащен пользователь из Optional, который возвращает метод findById, обработана возможность выбросить исключение, если в Optional ничего нет. И в итоге при вызове метода сервиса получение пользователя по id он вернёт мне чистого пользователя, или выбросить исключение. Дополнительную обработку делать не нужно.

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentAdminController.java:
```diff
@@ -0,0 +1,87 @@
+    private final CommentService service;
+
+    @Autowired
+    public CommentAdminController(CommentService service) {
```
Можно использовать аннотацию @RequiredArgsConstructor, она создаёт необходимый конструктор

EvgeniyaKuzmina replied:
Да, верно. Такая аннотация есть, и я её использовала. Но у меня почему-то эта аннотация не сработала в одном из сервисов, и падало с ошибкой всё, где было сказано, что необходим конструктор. Поэтому я во всём приложении использовала одинаковый подход — создание конструктора вручную.

Anton-Morozov-Java-3-2 replied:
Это странно

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentAdminController.java:
```diff
@@ -0,0 +1,87 @@
+                                                         @RequestParam(defaultValue = "desc") String sort,
+                                                         @RequestParam(defaultValue = FROM) @PositiveOrZero Integer from,
+                                                         @RequestParam(defaultValue = SIZE) @Positive Integer size) {
+        log.info("CommentAdminController: getAllCommentsByStatus — получен запрос на получение всех комментариев пользователя");
```
Мне кажется будет лучше, если логику из контроллеров перенести в сервисы.

EvgeniyaKuzmina replied:
Возможно я не так поняла, но логики в сервисах у меня вроде как нет. А эта строка кода — это просто логирование, чтобы можно было понять что запрос в принципе пришёл в контроллер, а не упал где-то по дороге.

Anton-Morozov-Java-3-2 replied:
Если есть ветвления if/else это считается логикой.

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentPrivateController.java:
```diff
@@ -0,0 +1,102 @@
+    @Autowired
+    public CommentPrivateController(CommentService service) {
+        this.service = service;
+    }
```
также @RequiredArgsConstructor

EvgeniyaKuzmina replied:
ответила выше

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentPrivateController.java:
```diff
@@ -0,0 +1,102 @@
+                                    @PathVariable Long userId) {
+        log.info("CommentPublicController: getAllCommentsByEventId — получен запрос на обновление комментария");
+        Comment comment = CommentMapper.fromUpdateCommentDto(commentDto);
+        comment = service.changeCommentByAuthor(comment, userId);
```
Лучше dto отправлять в service и возвращать в controller тоже dto, тогда вся логика работы с данными будет в service. В контроллерах должна остаться только валидация и интерфейс.

EvgeniyaKuzmina replied:
Думала об этом, и прошлую работу писала с таким подходом. Но как я поняла, всё же нет какого-то документа, в котором бы описывалось как правильно работать с ДТО. Скорее рекомендации к написанию. Но как мне кажется, в методах самих сервисов и так достаточно объёмные решения, и мне не хотелось нагружать их ещё и тем что формировать из ДТО обычные классы.

Anton-Morozov-Java-3-2 replied:
Хорошо, можешь оставить так

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/controller/CommentPrivateController.java:
```diff
@@ -0,0 +1,102 @@
+                                                           @RequestParam(defaultValue = "desc") String sort,
+                                                           @RequestParam(defaultValue = FROM) @PositiveOrZero Integer from,
+                                                           @RequestParam(defaultValue = SIZE) @Positive Integer size) {
+        log.info("CommentPublicController: getAllCommentsByEventId — получен запрос на получение всех комментариев пользователя");
```
Также перенести в сервис + тут есть дублирование кода из метода 

public Collection<CommentDto> getAllCommentsByStatus

это можно красиво оформить в методах проверки в слое сервисов.

EvgeniyaKuzmina replied:
Опять такие не поняла, почему нужно перенести в сервис обработку параметров, которые метод получает на вход.
В методе контроллера я проверяю, какие входные данные пришли мне на вход. Если какой-то параметр неверный — выбрасывается исключение. Зачем нести это в сервис и проверять там. Это лишнее действие, когда можно отсеять на входе.

Anton-Morozov-Java-3-2 replied:
Я про логику в if/else

EvgeniyaKuzmina replied:
Да, я поняла тебя :)
Но мне всё же кажется нелогично нести в сервис параметры, которые изначально могут быть с ошибкой.
Опять-таки я отталкиваюсь от того, что уже обсуждала ранее. Как я понимаю и мне кажется это логичным, в сервер всё же должны прийти уже чистые, проверенные данные

Anton-Morozov-Java-3-2 replied:
Да, но все равно ты делаешь проверки в слое сервера. Наверное это тоже рекомендации, если хочешь оставляй. Я думаю наставник на демо сможет прокомментировать этот вопрос.

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/dto/CommentDto.java:
```diff
@@ -1,29 +1,32 @@
-    private String creat;
+    private UserShortDto author;
+    private EventShortDtoForComment event;
+    private String created;
```
Можно использовать аннотацию @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss") и объявить дату в виде LocalDataTime

EvgeniyaKuzmina replied:
Спасибо большое за информацию, почему-то не подумала что может быть уже готовая такая аннотация. 
Буду знать на будущие проекты :)
В данном случае у меня во всём проекте используется форматирование через метод setter, поэтому оставлю пока так как есть.

Anton-Morozov-Java-3-2 replied:
Ок

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/dto/CommentShortDto.java:
```diff
@@ -0,0 +1,28 @@
+    private Long id;
+    private String text;
+    private UserShortDto author;
+    private String created;
```
также @JsonFormat(pattern = "yyyy-MM-dd HH:mm:ss")

EvgeniyaKuzmina replied:
ответила выше

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/mapper/CommentMapper.java:
```diff
@@ -0,0 +1,53 @@
+    private CommentMapper() {
+    }
+
+    public static CommentShortDto toCommentShortDto(Comment comment) {
```
необязательно использовать, но советую попробовать mapstruct как Mapper, очень удобная вещь

https://www.baeldung.com/mapstruct

EvgeniyaKuzmina replied:
Спасибо большое за совет. Будут знать

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/model/Comment.java:
```diff
@@ -35,5 +37,22 @@ public class Comment {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Comment comment = (Comment) o;
+        return text.equals(comment.text) && event.equals(comment.event) && author.equals(comment.author);
```
А почему сравниваются только три поля?

EvgeniyaKuzmina replied:
Идея была в том, что комментарий считается одинаковым если в нём идентичный текст, автор и событие к которому оно оставлено. Остальные поля могут быть не уникальными и могут повторяться в разных комментариях.

Anton-Morozov-Java-3-2 replied:
Лучше удалить это переопределение, могут быть ошибки или название поменять.

EvgeniyaKuzmina replied:
Удалила

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,185 @@
+        comment.setAuthor(user);
+        log.info("CommentServiceImpl: addNewComment — Комментарий добавлен {}.", comment);
+        return repository.save(comment);
+    }
```
Может ли пользователь оставлять несколько комментариев к событию? 
Можно ли оставлять комментарии до даты проведения события?

EvgeniyaKuzmina replied:
Да, ответ на каждый вопрос. Задумка комментариев в том, что это именно комментарии к событию, а не отзывы. 
Участники могут общаться по какому событию, в котором они участвуют. Например, встреча собачников района. В комментариях они могут обсуждать какие-то детали встречи, до самой встречи. 
Не хотелось бы ограничивать пользователя оставлять только один комментарий к событию.

Anton-Morozov-Java-3-2 replied:
Ок

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Override
+    public Comment changeCommentByAuthor(Comment updComment, Long userId) {
```
Нужно ли проверять актуальность заявки на участие при изменении комментария, у пользователя же есть возможность отменить свою одобренную заявку?

EvgeniyaKuzmina replied:
Упустила это момент. 
Добавлю проверку, спасибо

Anton-Morozov-Java-3-2 commented on ewm-service/src/main/java/ru/yandex/practicum/mainservice/event/comment/CommentServiceImpl.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Override
+    public void removeComment(Long commentId, Long userId) {
```
я думаю нужно сделать ограничение на удаление комментария (например автор события, комментатор, админ)?

EvgeniyaKuzmina replied:
Задумка в комментариях следующая:
Админ проверяет комментарии(модерирует их), и либо публикует комментарий — тогда его видят все пользователи, либо отклоняет комментарий — и его не видит никто, кроме самого создателя комментария.
Создатель в данном случае может удалить комментарий или изменить его и снова отправить на модерацию.
Удалять комментарии соответственно может только сам создатель. Владелец события не может удалить комментарий — да и зачем? Владелец события, как и другие пользователи, видит только опубликованные комментарии к его событию, то есть те комментарии, которые прошли модерацию.

Anton-Morozov-Java-3-2 replied:
А есть проверка что удаляющий владелец комментария?

EvgeniyaKuzmina replied:
да, есть

Anton-Morozov-Java-3-2 reviewed (commented):
Посмотрел работу. Всё классно, нужно провести небольшой рефакторинг и уточнить ограничения на добавление/обновление/удаление комментария.

EvgeniyaKuzmina committed:
fix: добавлена проверка при изменении комментария, что пользователь является участником события (заявка на участие в статусе CONFIRMED)
Изменена логика метода проверки, что пользователь является участником события
stile: удалён неиспользуемый метод; все методы в репозитории и сервисах запроса приведены к общему стилю приложения

EvgeniyaKuzmina committed:
fix: проверка кодировки

EvgeniyaKuzmina committed:
fix: проверка кодировки

EvgeniyaKuzmina committed:
fix: проверка кодировки

EvgeniyaKuzmina committed:
fix: проверка кодировки

EvgeniyaKuzmina committed:
style: перенесены логи на другую строку в методах

EvgeniyaKuzmina committed:
style: изменены комментарии в логах и выбрасываемых исключениях с кириллицы на латиницу

EvgeniyaKuzmina committed:
style: изменены комментарии в логах и выбрасываемых исключениях с кириллицы на латиницу

EvgeniyaKuzmina committed:
style: изменены комментарии в логах и выбрасываемых исключениях с кириллицы на латиницу

EvgeniyaKuzmina committed:
style: изменены комментарии в логах и выбрасываемых исключениях с кириллицы на латиницу

EvgeniyaKuzmina committed:
style: удалён лишний импорт
